packages/angular/ssr/src/routes/route-config.ts

```diff
@@ -0,0 +1,146 @@
+  CSR,
+
+  /** Static Site Generation (SSG) mode, where content is pre-rendered at build time and served as static files. */
+  SSG,
+}
+
+/**
+ * Fallback strategies for Static Site Generation (SSG) routes.
+ * @developerPreview
+ */
+export enum SSGFallback {
+  /** Use Server-Side Rendering (SSR) as the fallback for this route. */
+  SSR,
+
+  /** Use Client-Side Rendering (CSR) as the fallback for this route. */
+  CSR,
+
+  /** No fallback; Angular will not handle the response if the path is not pre-rendered. */
+  None,
+}
+
+/**
+ * Common interface for server routes, providing shared properties.
+ */
+export interface ServerRouteCommon {
+  /** The path associated with this route. */
+  path: string;
+
+  /** Optional additional headers to include in the response for this route. */
+  headers?: Record<string, string>;
+
+  /** Optional status code to return for this route. */
+  status?: number;
+}
+
+/**
+ * A server route that uses AppShell rendering mode.
+ */
+export interface ServerRouteAppShell extends Omit<ServerRouteCommon, 'headers' | 'status'> {
+  /** Specifies that the route uses AppShell rendering mode. */
+  renderMode: RenderMode.AppShell;
+}
+
+/**
+ * A server route that uses Client-Side Rendering (CSR) mode.
+ */
+export interface ServerRouteCSR extends ServerRouteCommon {
+  /** Specifies that the route uses Client-Side Rendering (CSR) mode. */
+  renderMode: RenderMode.CSR;
+}
+
+/**
+ * A server route that uses Static Site Generation (SSG) mode.
+ */
+export interface ServerRouteSSG extends Omit<ServerRouteCommon, 'status'> {
+  /** Specifies that the route uses Static Site Generation (SSG) mode. */
+  renderMode: RenderMode.SSG;
+
+  /**
+   * Optional fallback strategy to use if the SSG path is not pre-rendered.
+   * Defaults to `SSGFallback.SSR` if not provided.
+   */
+  fallback?: SSGFallback;
+  /**
+   * A function that returns a Promise resolving to an array of objects, each representing a route path with URL parameters.
+   * This function runs in the injector context, allowing access to Angular services and dependencies.
+   *
+   * @returns A Promise resolving to an array where each element is an object with string keys (representing URL parameter names)
+   * and string values (representing the corresponding values for those parameters in the route path).
+   *
+   * @example
+   * ```typescript
+   * export const serverRouteConfig: ServerRoutes[] = [
+   *   {
+   *     path: '/product/:id',
+   *     remderMode: RenderMode.SSG,
+   *     async getPrerenderPaths() {
+   *       const productService = inject(ProductService);
+   *       const ids = await productService.getIds(); // Assuming this returns ['1', '2', '3']
+   *
+   *       return ids.map(id => ({ id })); // Generates paths like: [{ id: '1' }, { id: '2' }, { id: '3' }]
+   *     },
+   *   },
+   * ];
+   * ```
+   */
+  getPrerenderPaths?: () => Promise<Record<string, string>[]>;
```

Comment:
Consider: Should this support `Iterable` / `AsyncIterable`? That would allow arrays to work but also potentially allow us to stream rendering (potentially as a future feature, doesn't necessarily need to be in scope for this PR).

Reply:
I'm not entirely clear on the point about streams here. We could include `Iterable` or `AsyncIterable`, but I don't think `AsyncIterable` is necessary in this context. It's also worth noting that, as of now, none of the Angular APIs use `AsyncIterable` as a type.

Reply:
My thinking was that this would leave open the door to potentially parallelizing the rendering process. For example:

```typescript
async function getPrerenderPaths(): AsyncIterable<Record<string, string>> {
  yield { foo: 'bar' };
  await someTime();
  yield { foo: 'baz' };
}
```

Angular can theoretically start prerendering the `{ foo: 'bar' }` case while waiting for `someTime`. Using `yield` can also just generally be a nice way of authoring this function, even if it doesn't parallelize anything.

Reply:
I agree that this is a good approach to authoring the function. However, since most users are likely to use `fetch` or `HttpClient` to populate data for the prerender paths, `AsyncIterable` might be somewhat redundant.

It's also important to note that the route tree is created before prerendering starts.

I’m open to adding `AsyncIterable` as a type, but at this stage, I don’t see a strong enough reason to expand the API. We can definitely reconsider and add it as a feature in the future.

Reply:
I'm ok with deferring this, that's potentially a feature we could add in a minor anyways, so I don't think we need `Iterable` / `AsyncIterable` to insulate from a future breaking change. Just a potential future optimization we can consider.